File: Data Structures & Algorithms/palindromic-substrings/submission-0.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        """
        - abcabcdabc
        - Traverse string 
        - A set to keep track of palindromic strings 
        - Keep two pointers as sliding windows. 
        - For each each if its a palindrome
        - If its a palindrome update longest
        - If not; then move pointer to the next char and do palindrome check
        - 
        """

        res = 0

        for i in range(len(s)):
            l, r = i, i
            # odd length palindrome
            while l >= 0 and r < len(s) and s[l] == s[r]:
                res += 1
                l -= 1
                r += 1


            # even length palindrome
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                res += 1
                l -= 1
                r += 1 
        
        return res





        

File: Data Structures & Algorithms/longest-palindromic-substring/submission-0.js
class Solution {
    /**
     * @param {string} s
     * @return {string}
     */
    longestPalindrome(s) {
        if (s.length <= 1) return s;

        let longest = "";

        const expandAroundCenter = (left, right) => {
            while (left >= 0 && right < s.length && s[left] === s[right]) {
                left--;
                right++;
            }
            // slice from (left + 1) to (right - 1), inclusive
            return s.slice(left + 1, right);
        };

        for (let i = 0; i < s.length; i++) {
            // Odd-length palindrome
            const odd = expandAroundCenter(i, i);
            if (odd.length > longest.length) longest = odd;

            // Even-length palindrome
            const even = expandAroundCenter(i, i + 1);
            if (even.length > longest.length) longest = even;
        }

        return longest;
    }
}
